src/pmhn/_trees/_simulate.py

```diff
@@ -1,23 +1,88 @@
-from typing import Union, Sequence
-
+from typing import Union, Sequence, Optional
+from anytree import Node
 import numpy as np
 
-from pmhn._trees._interfaces import Tree
-
 
-def _simulate_tree(
+def generate_valid_tree(
     rng,
     theta: np.ndarray,
     sampling_time: float,
-) -> Tree:
-    """Simulates a single tree with known sampling time.
+    mean_sampling_time: float,
+    min_tree_size: Optional[int] = None,
+    max_tree_size: Optional[int] = None,
+) -> tuple[dict[Node, float], float]:
+    """
+    Generates a single valid tree with known sampling time.
 
     Args:
         rng: random number generator
         theta: real-valued (i.e., log-theta) matrix,
           shape (n_mutations, n_mutations)
         sampling_time: known sampling time
+        min_tree_size: minimum size of the tree
+        max_tree_size: maximum size of the tree
+        mean_sampling_time: parameter for rng
+    Returns:
+        A valid mutation tree that meets the size constraints if specified.
+
+    Note:
+        The min_tree_size and max_tree_size parameters consider the entire tree,
+        i.e the root node is included.
+        To disable the size constraints, leave min_tree_size and max_tree_size as None.
+        If a tree is discarded, a new sampling time is drawn.
+    """
+
+    while True:
+        tree = _simulate_tree(rng, theta, sampling_time, max_tree_size)
+        if (min_tree_size is None or len(tree) >= min_tree_size) and (
+            max_tree_size is None or len(tree) <= max_tree_size
+        ):
+            return tree, sampling_time
+        else:
+            sampling_time = rng.exponential(scale=mean_sampling_time)
+
+
+def _find_possible_mutations(old_mutations: list[int], n_mutations: int) -> list[int]:
+    """
+    Args:
+        old_mutations: list of ancestor mutations of
+            a given node (including the node itself)
+        n_mutations: total number of mutations
+    Returns:
+        a list of possible mutations that could appear next for a given node
+
+    Note:
+        We assume that mutations are labeled with a number between
+        1 and n_mutations,
+        so each element in old_mutations should be in that range
+        (except for the root node = mutation 0).
+        If this assumption is violated, an exception is raised.
+
+    """
+    for mutation in old_mutations:
+        if mutation > n_mutations or mutation < 0:
+            raise ValueError(
+                f"Invalid mutation {mutation} in old_mutations."
+                f" It should be 0 <= mutation <= {n_mutations}."
+            )
+
+    possible_mutations = list(
+        set([i + 1 for i in range(n_mutations)]).difference(set(old_mutations))
```

Comment:
Consider `set(range(1, n_mutations + 1))` instead of `set([i + 1 for i in range(n_mutations)])`